mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,366 @@
+    num_training_circuits: int,
+    fraction_non_clifford: float,
+    method_select: str = 'random',
+    method_replace: str = 'nearest',
+    sigma_select: float = 0.5,
+    sigma_replace: float = 0.5,
+)-> list:
+
+    '''Returns a list of near-Clifford circuits to act as training data.
+
+    Args:
+        circuit: A circuit of interest.
+        num_training_circuits: Number of circuits in the returned training set, assumes already compiled into gate set
+                               (Rz, Rx, Z, X, CNOT)
+        fraction_non_clifford: The (approximate) fraction of non-Clifford gates in each returned circuit.
+        method_select: option to define the way in which the non-Clifford gates to replace with Cliffords are chosen.
+        method_replace: str = option to define the way the chosen non-Clifford gates are replace with a Clifford gate.
+        sigma_select: variable definined width of probability distribution used in choosing which non-Cliffords to replace.
+        sigma_replace: variable definined width of probability distribution used in choosing which Clifford to replace the non
+                       -Clifford with.
+    Returns:
+        List: [map_to_near_clifford(circuit, fraction_non_clifford, method, can_compile) for _ in range(num_training_circuits)]
+    '''
+    circuits_list = []
+    # First turn circuit into an data array which is easier to deal with, empty circuit is used to store qubit layout:
+    data, empty_circuit = circuit_to_array(circuit)
+    mask_rz = data[1, :] == 'rz'
+    rz_circ_data = data[:, mask_rz]
+    mask_not_rz = data[1, :] != 'rz'
+    not_rz_circ_data = data[:,mask_not_rz]
+    mask_non_cliff = is_clifford_angle(rz_circ_data[2, :]) == False
+    rz_non_cliff = rz_circ_data[:, mask_non_cliff]
+    mask_cliff = is_clifford_angle(rz_circ_data[2, :]) == True
+    rz_cliff = rz_circ_data[:, mask_cliff]
+    total_non_cliff = len(rz_non_cliff[0])
+    # define N:
+    N = int(fraction_non_clifford * total_non_cliff)
+
+    # find all the non-Clifford gates:
+    all_cliff = np.column_stack((not_rz_circ_data, rz_cliff))
+
+    for n in range(num_training_circuits):
+        empty_circuit_copy = empty_circuit.copy()
+        rz_non_cliff_copy = rz_non_cliff.copy()
+
+        # Choose non Clifford gates to change according to selection methods:
+        if method_select == 'random':
+            columns_to_change = sample(np.arange(0, total_non_cliff,1).tolist(), total_non_cliff-N)
+
+        elif method_select == 'probabilistic':
+            non_cliff_angles = rz_non_cliff_copy[2]
+            # form a probability distribution: 
+            probabilities = angle_to_probabilities(non_cliff_angles, sigma_select)
+            prob_choose_gate = [k / sum(probabilities) for k in probabilities]
+            columns_to_change = np.random.choice(np.arange(0, total_non_cliff,1).tolist(), total_non_cliff - N, replace=False,
+                                                p=prob_choose_gate)
```

Comment:
Factor this out into a function `map_to_near_clifford` which inputs a circuit and outputs one training (near clifford) circuit.